Pull common parts of skymatch from romancal and jwst into stcal

WilliamJamieson opened the pull request:
This PR pulls the common parts of skymatch from romancal and jwst into stcal to be shared.

Note, no tests are included because it is well tested by the step tests in both romancal and jwst.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

emolter commented on src/stcal/skymatch/skystatistics.py:
```diff
@@ -0,0 +1,136 @@
+from copy import deepcopy
+
+# THIRD PARTY
+from stsci.imagestats import ImageStats  # type: ignore[import-untyped]
```
might be worth considering making the relevant third-party type ignores global by putting them here https://github.com/spacetelescope/stcal/blob/d6a87c346b56daafcd6b8b43231fdbb8b9222965/pyproject.toml#L212

emolter commented on src/stcal/skymatch/skystatistics.py:
```diff
@@ -0,0 +1,136 @@
+
+class SkyStats:
+    """
+    This is a superclass build on top of
+    :py:class:`stsci.imagestats.ImageStats`. Compared to
```
This comment is confusing because the class doesn't actually inherit from `ImageStats`

mcara replied:
Docs could be changed to explain that `SkyStats` delegates calls to an `ImageStats` object if that is more clear but the comment about persistence of settings is important as it is what differentiates it from `ImageStats` and I think it helps with performance.

emolter replied:
I agree that the comment itself explains an important distinction, and yes the simple wording change from "superclass built on top of" to "delegates calls to" would be clearer

WilliamJamieson commented:
> This looks great, thanks William! I just had a few questions about whether some of the ruff and mypy problems could be fixed instead of ignored.
> 
> I also had a few questions about the code itself, but those also exist on jwst main, so they're technically beyond the scope of this PR. But I'm still curious to know the answers to those and I figure now is as good a time as any to ensure the code is written as cleanly as possible

I agree they should be fixed. However, as you said they are beyond the scope of this PR. I think for making it possible to trace the history of changes it makes sense to move the code in as close a state as possible from JWST to stcal and then do a follow on PR making cleanups and addressing the issues uncovered here.

kmacdonald-stsci commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,1020 @@
+
+        # "bottom" points:
+        borderx[:nptx] = xs
+        bordery[:nptx] = -0.5
```
I suggest a comment explaining the -0.5 number or the use of a descriptive constant detailing what this number is.

WilliamJamieson replied:
I agree, and I believe this has something to do with all the pixel offsetting. However this is largely just a copy of the stuff from JWST with minimal changes. If someone knows exactly what this is I would be happy to add a comment here.

Really, we should be maintaining a common "constants" object/module to document these and back reference that as opposed to having a bunch of ad hoc magic numbers some of which are documented with comments. This however, is something to consider later.

mcara replied:
Centers of pixels have integer image coordinates. Edges of the pixels are located `+/- 0.5` pixels away from centers. I think this convention is explained somewhere in the docs. IMO, it does not make sense to add a comment like this every single time we see a 0.5 in the code.

mcara replied:
and `-0.5` is shorter than `coordinate_of_left_pixel_border(_relative_to_center)`

emolter replied:
I agree with Mihai that in the context of pixels it's relatively clear what a 0.5 offset is doing. 
 However it looks to me like `calc_bounding_polygon` is the only function where this is really relevant, so would it work to add a note to the docstring saying something like,
"""
Notes
-----
The bounding polygon is defined from corners of pixels
whereas the reference pixel is defined at the center of a pixel
and therefore the lower-left corner is located at (-0.5, -0.5)
"""

or whatever you want, I'm sure I made some small inaccuracies in describing that but you get the idea

mcara replied:
"whereas the reference pixel" (in @emolter suggestion) should be something like "whereas the pixel coordinates refer to their centers" (or something like that). "Reference pixel" has a special meaning in the WCS world.

emolter replied:
I see that this was resolved but I don't see any docstring update, is there any objection to my previous suggestion?

WilliamJamieson replied:
Oops I copied over that with something, I'll re-add this

emolter replied:
thanks!

emolter commented:
> I agree they should be fixed. However, as you said they are beyond the scope of this PR.

IMO making the code adhere to the more stringent style rules in stcal remains within the scope.  I don't much like ignoring them.  I think the downstream testing should assuage any concerns that those changes might break something.

I agree that my comment RE the potentially unnecessary use of `@property` can be ignored for now though.

mcara commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,1020 @@
+        pix_area=1.0,
+        convf=1.0,
+        mask=None,
+        id=None,  # noqa: A002
```
probably could be changed to `image_id`???

mcara replied:
I suggest `sky_id` as it would work with both `SkyImage` and `SkyGroup`.

mcara commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,1020 @@
+
+            .. note::
+
+              The functionality to support this conversion is not yet
+              implemented and at this moment `convf` is ignored.
```
Based on ancient (>9 years old) discussions with relevant people, this feature, _in theory_, may not be needed for JWST. _In practice_, however, we don't know. I suggest leaving this alone as is.

WilliamJamieson replied:
This is actually being used in places

emolter replied:
Can you please point me to where?  If that's the case, this comment in the docstring should be removed

WilliamJamieson replied:
Thats all logging but I'm not qualified to tell if we want to keep any parts of those logs

emolter replied:
I don't think "used in logging" qualifies as being used given that `convf` is hard-coded to equal 1.0 and therefore has no effect on the scaling reported in the log messages.  So I agree with Mihai we should leave this as-is, including the docstring

mcara commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,1020 @@
+            self._image_shape = self._image.get_data_shape()
+        return self._image_shape
+
+    @property
+    def id(self):  # noqa: A003
```
Since the original code was written ~11-12 years ago (not sure about this but I think so), I no longer remember the intention. Maybe during the development I thought that, for example, `pix_area` will be computed on demand but ended up storing it for efficiency. It just allows more flexibility. Also, since both `SkyImage` and `SkyGroup` can be used as inputs to `match`, they should have  a common set of properties. I do not remember what was the Python version used during the development but it may have been `2.x` (due to PyRAF / Tables/ whatever limitations) so I don't think `ABC` was available at that time and even if it was, I am not sure properties or attributes of classes could be made abstract. Another reason may have been the desire to get these properties documented (in the docstrings).

I do not think anything should be changed in this PR but in the future they could subclass from a `BaseSky`, `ABCSky` or something like that.

I do not see what is the harm in using properties.

emolter replied:
no harm it's just unnecessarily verbose since neither the getter nor the setter does anything.  It's therefore confusing, because when I see those methods, I expect them to do something.  But it doesn't matter- I agree with William's comment that these changes shouldn't go into this PR, if they happen at all.

mcara commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,1020 @@
+
+    """
+
+    def __init__(self, images, id=None, sky=0.0):  # noqa: A002
```
I suggest `sky_id` as it would work with both `SkyImage` and `SkyGroup`.

mcara commented on src/stcal/skymatch/skymatch.py:
```diff
@@ -0,0 +1,539 @@
+log.setLevel(logging.DEBUG)
+
+
+def match(images, skymethod="global+match", match_down=True, subtract=False):
```
I would suggest something along the lines `do_match` or `perform_match`, or `match_sky`. The intent for the name of this function was for it to be a verb. `skymatch` does not sound to me (a non-native English speaker) as a verb but I may be wrong.

emolter replied:
I'm happy with whatever name

emolter commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,946 @@
+            binwidth=binwidth
+        )
+
+    # TODO: due to a bug in the sphere package, see
+    #       https://github.com/spacetelescope/sphere/issues/74
+    #       intersections with polygons formed as union does not work.
+    #       For this reason I re-implement 'calc_sky' below with
+    #       a workaround for the bug.
+    #       The original implementation (now called ``_calc_sky_orig``
+    #       should replace current 'calc_sky' once the bug is fixed.
+    #
```
Can this be moved into a Note for the `calc_sky` function?

emolter commented on src/stcal/skymatch/skyimage.py:
```diff
@@ -0,0 +1,946 @@
+
+        return skyval, npix, polyarea
+
+#     def _calc_sky_orig(self, overlap=None, delta=True):
```
remove instead of commenting out

emolter reviewed (changes requested):
Thanks for all the updates!  Before merging I think it would be good to update https://github.com/spacetelescope/jwst/pull/8901 such that we can see if the changes are likely to work well with jwst.  Perhaps you were already planning to do so.

WilliamJamieson commented:
@emolter I believe this is ready now.

WilliamJamieson commented on docs/conf.py:
```diff
@@ -39,9 +39,16 @@
+nitpick_ignore = [
+    ("py:class", "optional"),
+    ("py:class", "np.ndarray"),
+    ("py:class", "stsci.imagestats.ImageStats"),  # intersphinx isn't working here
+    ("py:class", "spherical_geometry.polygon.SphericalPolygon"), # intersphinx isn't working here
```
These should be working with the intersphinx links I added. I believe the fix may require some fixes in the documentation for those packages.

It is possible something is off in the `stcal` docs causing the link failure. Please feel free to push changes to attempt to fix this.

emolter reviewed (commented):
All my previous questions have been answered now.

I don't think the changes have been so large that it would be helpful for me to go through it again, but let me know if you'd like me to.

So this looks good to me pending fixing the bug that is causing the failing jwst regtests

mcara commented:
It would still be useful to have some unit tests developed for this. We have always relied on regression tests (both for HST and JWST) but PRs to this code in the future should be able to use some quick unit tests. It could be done either as part of this PR or as a follow-up PR.

WilliamJamieson commented:
> It would still be useful to have some unit tests developed for this. We have always relied on regression tests (both for HST and JWST) but PRs to this code in the future should be able to use some quick unit tests. It could be done either as part of this PR or as a follow-up PR.

I think this should be a follow on PR. I looked into moving some of the unit tests from JWST. But at this time all those tests directly involve running the entire skymatch step. I personally don't feel like I can write any useful unit tests for this part of skymatch independent of JWST at this time.

emolter commented:
Thanks @nden, that's fine with me. As I said, I'll only look at this again if needed, @WilliamJamieson I see you re-requested my review after @nden dismissed the stale one.  Does that mean you'd like me to look again, or would you just like me to hit the Approve button?

WilliamJamieson commented:
> Thanks @nden, that's fine with me. As I said, I'll only look at this again if needed, @WilliamJamieson I see you re-requested my review after @nden dismissed the stale one. Does that mean you'd like me to look again, or would you just like me to hit the Approve button?

If you want to do another full review your welcome, though there hasn't been any substantive changes since I addressed your last one. In any case, in order for me to merge this PR I do need two approving reviews.

nden commented:
@schlafly @ddavis-stsci This is ready to be merged from JWSt point of view. Romancal unit tests are passing and a [PR using this code is in the works for Roman.](https://github.com/spacetelescope/romancal/pull/1465). Please review when available.

tapastro commented:
This was waiting for a Roman team member to approve - in general, stcal PRs require approvals not just from any two people, but one from jwst and one from Roman. This still needs eyes from the Roman team. 

We can make this procedure more clear to the branch.

WilliamJamieson commented:
> This was waiting for a Roman team member to approve - in general, stcal PRs require approvals not just from any two people, but one from jwst and one from Roman. This still needs eyes from the Roman team.
> 
> We can make this procedure more clear to the branch.

I have been tracking this for both JWST and RCAL. JWST was the main sticking point with RCAL waiting for this to go in.

tapastro commented:
Eddie, the Romancal product owner, listed himself and Dave Davis as approved reviewers for stcal PRs. He is free to add you to that list, but I have not heard of an update there and am trying to follow protocol so that we do not merge changes that Romancal is not aware of.

schlafly commented:
Thanks.  I'm happy for this to go in.

I'll note that I got confused trying to understand how the "self.sky" feature was supposed to function in the SkyGroups, and where or how the image.backgrounds are used.
https://github.com/spacetelescope/stcal/pull/310/files#diff-1d5ed0c79f81a6487ae90d08de3bf1b263bb11b97aa1aee2175b28918a78f7b3R731
https://github.com/spacetelescope/stcal/pull/310/files#diff-1d5ed0c79f81a6487ae90d08de3bf1b263bb11b97aa1aee2175b28918a78f7b3R813

I'm sure I just missed it, though, and presumably it is duplicating existing code.
